### DISCUSSION TITLE
Can't annotate both int and smallint

### DISCUSSION THREAD

**Discussion by pdc1** [sqlalchemy/sqlalchemy#10659] [Usage Questions] [answered]
I am trying to use 2.0 type annotations, but I can't get both integer and smallint to work together. This is for v2.0.23 with a MariaDB database and the mariadb connector.

I have this (partial) definition:
```python
# Set up type aliases
small_int = int

class Base(DeclarativeBase):
    """SQLAlchemy base class."""

    type_annotation_map: ClassVar = {
        small_int: SmallInteger(),
    }

class BookDB(MappedAsDataclass, Base):
    """Model for book."""
    __tablename__ = "book"
    [...]
    gr_id: Mapped[Optional[int]]
    year_published: Mapped[Optional[small_int]]
```

Which results in both fields being `SMALLINT`:
```sql
CREATE TABLE book (
        [...]
        gr_id SMALLINT, 
        year_published SMALLINT, 
        [...]
)
```

I tried adding a definition for `integer = int`, but that results in both fields being either `INTEGER` or `SMALLINT`, depending on the order of `integer` and `small_int` (or perhaps the order in the `type_annotation_map`). 

I can work around this with an explicit `mapped_column(Integer())` mapping, but I'd like to use the type annotations if possible.

Thanks!

**Comment by CaselIT** [comment, marked as answer]
Hi,

Well thats expected, the mapping is a dict so you can't have the same object twice in it.

You may want to look into the `Annotated` support

**Reply by pdc1** [reply]
Just so I understand what is happening, are you saying that since `small_int` is defined to be an `int`, that SQL Alchemy type mapping can't tell the difference between the two `int` types?

I tried changing `small_int` to be `Annotated[int, 6]` and that works. (I chose 6 since that's how many digits are in a small_int, I'm not sure if there is a good "no-op" to use with `Annotated`).

Thanks!

**Reply by CaselIT** [reply]
> Just so I understand what is happening, are you saying that since small_int is defined to be an int, that SQL Alchemy type mapping can't tell the difference between the two int types?

yes, if you do `small_int is int` you get `True`. Same if you try to add it to a set or dict.

> I tried changing small_int to be Annotated[int, 6] and that works. (I chose 6 since that's how many digits are in a small_int, I'm not sure if there is a good "no-op" to use with Annotated).

You could also do `small_int = Annotated[int, mapped_column(SmallInteger)]`. This way you could avoid setting it to the type map. But what you did works too, the important thing is that sqlalchemy can tell it apart from int

**Reply by CaselIT** [reply]
this is documented here. There are various examples https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#mapping-multiple-type-configurations-to-python-types

**Reply by pdc1** [reply]
Awesome, thanks! I considered the `mapped_column` variant, but I have other type mappings for strings and whatnot, so a minimal Annotation fit better with my current definitions.
